### ISSUE TITLE
Multi-GPU bugs under windows

### ISSUE THREAD

**Issue by ommiissyu** [dmlc/mxnet#295]
hi, all

I have tried mxnet under windows. When I use 1 GPU cards, it works well, but when using multi-gpus, it   will crush, the trace message is like this:
```
Traceback (most recent call last):
  File "D:\users\bixi\Source\mxnet\example\mimagenet\inception.py", line 101, in <module>
    iter_end_callback=mx.callback.do_checkpoint(model_prefix))
  File "C:\Python27\lib\site-packages\mxnet-0.5.0-py2.7.egg\mxnet\model.py", line 677, in fit
    logger=logger)
  File "C:\Python27\lib\site-packages\mxnet-0.5.0-py2.7.egg\mxnet\model.py", line 292, in _train_mul
ti_device
    kv.push(index, grad_list, priority=-index)
  File "C:\Python27\lib\site-packages\mxnet-0.5.0-py2.7.egg\mxnet\kvstore.py", line 170, in push
    ctypes.c_int(priority)))
WindowsError: exception: access violation reading 0xFFFFFFFFFFFFFFFF
```

**Comment by jh-yu** [comment]
It seems the problem is in the function `MXKVStoreSetUpdater`(c_api.cc, line 1003). When the lambda function is called, the `updater`'s access violates. I am not sure why windows behaves like this(maybe it is because the memory space of `updater` is across the edge between python and c++?). But you could simply add a line at line 1003 after `API_BEGIN();`,  `auto win_updater = std::move(updater);` and then use `win_updater` to replace `updater` in line 1008.
It works for me.

**Comment by yajiedesign** [comment]
i am fix it.

**Comment by tqchen** [comment]
Fixed by #299

**Comment by tqchen** [comment]
This was caused by lambda capture of function type instead of function pointer type. 

@mli We might need to be a bit careful about it in future, we could consider change the C API to accept the function pointer type instead.
